[Nano] How-To Guides: Training - TensorFlow

Oscilloscope98 opened the pull request:
## Description

### 1. Why the change?

Add Nano how-to guides to our documentation for better user experience. How-to guides aim at giving multiple bite-sized, task-oriented, and executable examples for user to check when they need to do similar tasks.

### 2.  Summary of the change 

In this PR, guides for Training - TensorFlow Keras is covered. The guides source files are located under folder BigDL/python/nano/tutorial/notebook/training/tensorflow.

- [x] [How to accelerate a TensorFlow Keras application on training workloads through multiple instances](https://yuwentestdocs.readthedocs.io/en/nano-howto-training-tf/doc/Nano/Howto/Training/TensorFlow/accelerate_tensorflow_training_multi_instance.html) (based on [this](https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/training/tensorflow/tensorflow_train_multi_instance.py))
- [x] [How to optimize your model with a sparse Embedding layer and SparseAdam optimizer](https://yuwentestdocs.readthedocs.io/en/nano-howto-training-tf/doc/Nano/Howto/Training/TensorFlow/tensorflow_training_embedding_sparseadam.html) (based on [this](https://github.com/intel-analytics/BigDL/blob/main/python/nano/tutorial/training/tensorflow/tensorflow_sparse_embedding.py))

And non-runnable guide:
- [x] [How to choose the number of processes for multi-instance training](https://yuwentestdocs.readthedocs.io/en/nano-howto-training-tf/doc/Nano/Howto/Training/General/choose_num_processes_training.html)

### 3. How to test?
- [x] Document test: https://yuwentestdocs.readthedocs.io/en/nano-howto-training-tf/doc/Nano/Howto/index.html#tensorflow
- [x] GitHub Notebook Preview
- [How to accelerate a TensorFlow Keras application on training workloads through multiple instances](https://github.com/Oscilloscope98/BigDL/blob/nano-howto-training-tf/python/nano/tutorial/notebook/training/tensorflow/accelerate_tensorflow_training_multi_instance.ipynb)
- [How to optimize your model with a sparse Embedding layer and SparseAdam optimizer](https://github.com/Oscilloscope98/BigDL/blob/nano-howto-training-tf/python/nano/tutorial/notebook/training/tensorflow/tensorflow_training_embedding_sparseadam.ipynb)
- [x] Notebook test on github action

Notebook test locally (conda create an empty environment with python=3.7)
- [x] How to accelerate a TensorFlow Keras application on training workloads through multiple instances
- [x] How to optimize your model with a sparse Embedding layer and SparseAdam optimizer

TheaperDeng commented:
> When importing `Model` from `bigdl.nano.tf.keras` after creating datasets etc., error `Inter op parallelism cannot be modified after initialization` may occur (for example in the codes [here](https://github.com/intel-analytics/BigDL/blob/dd559a48199219eb9b9ee27c10dccf70d5140562/python/nano/tutorial/notebook/training/tensorflow/accelerate_tensorflow_training_multi_instance.ipynb)): ![MicrosoftTeams-image](https://user-images.githubusercontent.com/54161268/191145714-7186c57a-d084-4db3-8be7-a55eeec5587d.png) @TheaperDeng Is this something we should fix? Or is this a limitation of `bigdl.nano.tf.keras.Model` we should mention in the how-to guide?

I think we have met similar issue when we write tf based chronos model right? how did we resolved it? @liangs6212

Oscilloscope98 commented on python/nano/tutorial/notebook/training/tensorflow/run-nano-howto-guides-training-tensorflow-tests.sh:
```diff
@@ -0,0 +1,28 @@
+start=$(date "+%s")
+
+# use nbconvert to test here; nbmake may cause some errors
+jupyter nbconvert --ExecutePreprocessor.timeout=600 --to notebook --execute ${NANO_HOWTO_GUIDES_TEST_DIR}/*.ipynb
```
`jupyter nbconvert` is used here instead of `python -m pytest -s --nbmake --nbmake-timeout=600 --nbmake-kernel=python3 ${NANO_HOWTO_GUIDES_TEST_DIR}/*.ipynb` to avoid errors shown [here](https://github.com/intel-analytics/BigDL/actions/runs/3087422563/jobs/4992773410#step:6:1470):
![image](https://user-images.githubusercontent.com/54161268/191170020-d47ec7ad-4d05-4230-9a4c-f4dde02fc5f8.png)
![image](https://user-images.githubusercontent.com/54161268/191170088-0f7cd5bc-9ede-47f4-b826-851488619e2c.png)

liangs6212 commented:
> > When importing `Model` from `bigdl.nano.tf.keras` after creating datasets etc., error `Inter op parallelism cannot be modified after initialization` may occur (for example in the codes [here](https://github.com/intel-analytics/BigDL/blob/dd559a48199219eb9b9ee27c10dccf70d5140562/python/nano/tutorial/notebook/training/tensorflow/accelerate_tensorflow_training_multi_instance.ipynb)): ![MicrosoftTeams-image](https://user-images.githubusercontent.com/54161268/191145714-7186c57a-d084-4db3-8be7-a55eeec5587d.png) @TheaperDeng Is this something we should fix? Or is this a limitation of `bigdl.nano.tf.keras.Model` we should mention in the how-to guide?
> 
> I think we have met similar issue when we write tf based chronos model right? how did we resolved it? @liangs6212

I think it can be solved by modifying `define_model_inputs_outputs` to `define_model`.
like this: 
```python
def define_model(...):
    inputs = tf.keras.layers.Input(shape=(img_size, img_size, 3))
    x = tf.cast(inputs, tf.float32)
    x = tf.keras.applications.resnet50.preprocess_input(x)
    backbone = ResNet50(weights='imagenet')
    backbone.trainable = False
    x = backbone(x)
    x = layers.Dense(512, activation='relu')(x)
    outputs = layers.Dense(num_classes, activation='softmax')(x)
    model = Model(inputs=inputs, outputs=outputs)
    model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=['accuracy'])
    return model
```

Oscilloscope98 commented:
> I think it can be solved by modifying `define_model_inputs_outputs` to `define_model`. like this:
> 
> ```python
> def define_model(...):
>     inputs = tf.keras.layers.Input(shape=(img_size, img_size, 3))
>     x = tf.cast(inputs, tf.float32)
>     x = tf.keras.applications.resnet50.preprocess_input(x)
>     backbone = ResNet50(weights='imagenet')
>     backbone.trainable = False
>     x = backbone(x)
>     x = layers.Dense(512, activation='relu')(x)
>     outputs = layers.Dense(num_classes, activation='softmax')(x)
>     model = Model(inputs=inputs, outputs=outputs)
>     model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=['accuracy'])
>     return model
> ```

It seems that this cannot solve the problem. The problem occurred when importing `model` from `bigdl.nano.tf.keras` after we have importing `tensorflow` for creating datasets, like this:
```python
import tensorflow as tf
import tensorflow_datasets as tfds

def create_datasets(img_size, batch_size):
    (train_ds, test_ds), info = tfds.load('imagenette/320px-v2',
                                          data_dir='/tmp/data',
                                          split=['train', 'validation'],
                                          with_info=True,
                                          as_supervised=True)
    
    num_classes = info.features['label'].num_classes
    
    def preprocessing(img, label):
        return tf.image.resize(img, (img_size, img_size)), \
               tf.one_hot(label, num_classes)

    train_ds = train_ds.repeat().map(preprocessing).batch(batch_size)
    test_ds = test_ds.map(preprocessing).batch(batch_size)
    return train_ds, test_ds, info

train_ds, test_ds, ds_info = create_datasets(img_size=224, batch_size=32)

from bigdl.nano.tf.keras import Model # <= error occurs here
```
I currently implemented a solution to this. That is importing `Model` from `bigdl.nano.tf.keras` at the beginning of the code snippet like this:
```python
import tensorflow as tf
import tensorflow_datasets as tfds

from bigdl.nano.tf.keras import Model # <= import here

def create_datasets(img_size, batch_size):
    (train_ds, test_ds), info = tfds.load('imagenette/320px-v2',
                                          data_dir='/tmp/data',
                                          split=['train', 'validation'],
                                          with_info=True,
                                          as_supervised=True)
    
    num_classes = info.features['label'].num_classes
    
    def preprocessing(img, label):
        return tf.image.resize(img, (img_size, img_size)), \
               tf.one_hot(label, num_classes)

    train_ds = train_ds.repeat().map(preprocessing).batch(batch_size)
    test_ds = test_ds.map(preprocessing).batch(batch_size)
    return train_ds, test_ds, info

train_ds, test_ds, ds_info = create_datasets(img_size=224, batch_size=32)
```

What I am confused here is: Do we need to remind user of this import order to avoid error? Or is this error something we should fix for `bigdl.nano.tf.keras.Model`? @TheaperDeng

jason-dai commented:
> > I think it can be solved by modifying `define_model_inputs_outputs` to `define_model`. like this:
> > ```python
> > def define_model(...):
> >     inputs = tf.keras.layers.Input(shape=(img_size, img_size, 3))
> >     x = tf.cast(inputs, tf.float32)
> >     x = tf.keras.applications.resnet50.preprocess_input(x)
> >     backbone = ResNet50(weights='imagenet')
> >     backbone.trainable = False
> >     x = backbone(x)
> >     x = layers.Dense(512, activation='relu')(x)
> >     outputs = layers.Dense(num_classes, activation='softmax')(x)
> >     model = Model(inputs=inputs, outputs=outputs)
> >     model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=['accuracy'])
> >     return model
> > ```
> 
> It seems that this cannot solve the problem. The problem occurred when importing `Model` from `bigdl.nano.tf.keras` after we have importing `tensorflow` for creating datasets, like this:
> 
> ```python
> import tensorflow as tf
> import tensorflow_datasets as tfds
> 
> def create_datasets(img_size, batch_size):
>     (train_ds, test_ds), info = tfds.load('imagenette/320px-v2',
>                                           data_dir='/tmp/data',
>                                           split=['train', 'validation'],
>                                           with_info=True,
>                                           as_supervised=True)
>     
>     num_classes = info.features['label'].num_classes
>     
>     def preprocessing(img, label):
>         return tf.image.resize(img, (img_size, img_size)), \
>                tf.one_hot(label, num_classes)
> 
>     train_ds = train_ds.repeat().map(preprocessing).batch(batch_size)
>     test_ds = test_ds.map(preprocessing).batch(batch_size)
>     return train_ds, test_ds, info
> 
> train_ds, test_ds, ds_info = create_datasets(img_size=224, batch_size=32)
> 
> from bigdl.nano.tf.keras import Model # <= error occurs here
> ```
> 
> I currently implemented a solution to this. That is importing `Model` from `bigdl.nano.tf.keras` at the beginning of the code snippet like this:
> 
> ```python
> import tensorflow as tf
> import tensorflow_datasets as tfds
> 
> from bigdl.nano.tf.keras import Model # <= import here
> 
> def create_datasets(img_size, batch_size):
>     (train_ds, test_ds), info = tfds.load('imagenette/320px-v2',
>                                           data_dir='/tmp/data',
>                                           split=['train', 'validation'],
>                                           with_info=True,
>                                           as_supervised=True)
>     
>     num_classes = info.features['label'].num_classes
>     
>     def preprocessing(img, label):
>         return tf.image.resize(img, (img_size, img_size)), \
>                tf.one_hot(label, num_classes)
> 
>     train_ds = train_ds.repeat().map(preprocessing).batch(batch_size)
>     test_ds = test_ds.map(preprocessing).batch(batch_size)
>     return train_ds, test_ds, info
> 
> train_ds, test_ds, ds_info = create_datasets(img_size=224, batch_size=32)
> ```
> 
> What I am confused here is: Do we need to remind user of this import order to avoid error? Or is this error something we should fix for `bigdl.nano.tf.keras.Model`? @TheaperDeng

I think we set up Inter op parallelism when importing nano, which must be done before user importing any tensorflow code. I think we may need to introduce something like `patch_sklearn()` in sk-learn extension? @TheaperDeng @yangw1234

Oscilloscope98 commented on python/nano/tutorial/notebook/training/tensorflow/tensorflow_training_embedding_sparseadam.ipynb:
```diff
@@ -0,0 +1,344 @@
+    ">\n",
+    "> If you would like to apply a regularizer function to the embedding matrix through setting `embeddings_regularizer`, Nano will apply the regularizer to the output tensors of the embedding layer instead to avoid making the sparse gradient dense (if `activity_regularize=None`).\n",
+    ">\n",
+    "> Please refer to [here](https://github.com/intel-analytics/BigDL/blob/main/python/nano/src/bigdl/nano/tf/keras/layers/embeddings.py) for more information on `bigdl.nano.tf.keras.layers.Embedding`."
```
I provide link to our source code here for user to get more info on `bigdl.nano.tf.keras.layers.Embedding`, and also the source code link for `bigdl.nano.tf.optimizers.SparseAdam` later. Because we do not have related api docs for now.

Should we add related api docs for these two apis?

yangw1234 commented:
> > > I think it can be solved by modifying `define_model_inputs_outputs` to `define_model`. like this:
> > > ```python
> > > def define_model(...):
> > >     inputs = tf.keras.layers.Input(shape=(img_size, img_size, 3))
> > >     x = tf.cast(inputs, tf.float32)
> > >     x = tf.keras.applications.resnet50.preprocess_input(x)
> > >     backbone = ResNet50(weights='imagenet')
> > >     backbone.trainable = False
> > >     x = backbone(x)
> > >     x = layers.Dense(512, activation='relu')(x)
> > >     outputs = layers.Dense(num_classes, activation='softmax')(x)
> > >     model = Model(inputs=inputs, outputs=outputs)
> > >     model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=['accuracy'])
> > >     return model
> > > ```
> > 
> > 
> > It seems that this cannot solve the problem. The problem occurred when importing `Model` from `bigdl.nano.tf.keras` after we have importing `tensorflow` for creating datasets, like this:
> > ```python
> > import tensorflow as tf
> > import tensorflow_datasets as tfds
> > 
> > def create_datasets(img_size, batch_size):
> >     (train_ds, test_ds), info = tfds.load('imagenette/320px-v2',
> >                                           data_dir='/tmp/data',
> >                                           split=['train', 'validation'],
> >                                           with_info=True,
> >                                           as_supervised=True)
> >     
> >     num_classes = info.features['label'].num_classes
> >     
> >     def preprocessing(img, label):
> >         return tf.image.resize(img, (img_size, img_size)), \
> >                tf.one_hot(label, num_classes)
> > 
> >     train_ds = train_ds.repeat().map(preprocessing).batch(batch_size)
> >     test_ds = test_ds.map(preprocessing).batch(batch_size)
> >     return train_ds, test_ds, info
> > 
> > train_ds, test_ds, ds_info = create_datasets(img_size=224, batch_size=32)
> > 
> > from bigdl.nano.tf.keras import Model # <= error occurs here
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I currently implemented a solution to this. That is importing `Model` from `bigdl.nano.tf.keras` at the beginning of the code snippet like this:
> > ```python
> > import tensorflow as tf
> > import tensorflow_datasets as tfds
> > 
> > from bigdl.nano.tf.keras import Model # <= import here
> > 
> > def create_datasets(img_size, batch_size):
> >     (train_ds, test_ds), info = tfds.load('imagenette/320px-v2',
> >                                           data_dir='/tmp/data',
> >                                           split=['train', 'validation'],
> >                                           with_info=True,
> >                                           as_supervised=True)
> >     
> >     num_classes = info.features['label'].num_classes
> >     
> >     def preprocessing(img, label):
> >         return tf.image.resize(img, (img_size, img_size)), \
> >                tf.one_hot(label, num_classes)
> > 
> >     train_ds = train_ds.repeat().map(preprocessing).batch(batch_size)
> >     test_ds = test_ds.map(preprocessing).batch(batch_size)
> >     return train_ds, test_ds, info
> > 
> > train_ds, test_ds, ds_info = create_datasets(img_size=224, batch_size=32)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What I am confused here is: Do we need to remind user of this import order to avoid error? Or is this error something we should fix for `bigdl.nano.tf.keras.Model`? @TheaperDeng
> 
> I think we set up Inter op parallelism when importing nano, which must be done before user importing any tensorflow code. I think we may need to introduce something like `patch_sklearn()` in sk-learn extension? @TheaperDeng @yangw1234

This looks like a tensorflow problem. You cannot set threads after running `tfds.load`. Not sure how to work around it yet.

```python
import tensorflow as tf
import tensorflow_datasets as tfds

(train_ds, test_ds), info = tfds.load('mnist',
                                      data_dir='/tmp/data',
                                      split=['train', 'test'],
                                      with_info=True,
                                      as_supervised=True)

print(tf.config.threading.get_inter_op_parallelism_threads()) # returns 0
tf.config.threading.set_inter_op_parallelism_threads(1)
```

shane-huang commented on docs/readthedocs/source/doc/Nano/Howto/Training/TensorFlow/accelerate_tensorflow_training_multi_instance.nblink:
```diff
@@ -0,0 +1,3 @@
+{
```
BigDL-Nano provides ``bigdl.nano.tf.keras.Model`` and ``bigdl.nano.tf.keras.Sequential`` which extend ```tf.keras.Model``` and ```tf.keras.Sequential``` respectively with various optimizations.  To use multi-instance training on a server with multiple CPU cores or sockets, you just replace `tf.keras.Model/Sequential` in your code with `bigdl.nano.tf.keras.Model/Sequential`, and call ``fit`` with specified num_processes.

shane-huang replied:
Do we have any tips about setting num_processes? If we have we can add them here, as a Related Reading or a Note.

Oscilloscope98 replied:
Have added a guide about how to choose number of processes for multi-instance training. @TheaperDeng Please have a look, thank you! https://yuwentestdocs.readthedocs.io/en/nano-howto-training-tf/doc/Nano/Howto/Training/General/choose_num_processes_training.html

TheaperDeng replied:
the guide looks good to me

shane-huang commented on python/nano/tutorial/notebook/training/tensorflow/accelerate_tensorflow_training_multi_instance.ipynb:
```diff
@@ -0,0 +1,255 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To benefit from multi-instance training, you need to **import** `Model` **or** `Sequential` **from** `bigdl.nano.tf.keras` **instead of** `tf.keras` **first**. Let's take the `Model` class here as an example:"
```
First, import Model or Sequential from bigdl.nano.tf.keras instead of tf.keras first. Let’s take the Model class here as an example:

Oscilloscope98 replied:
Have revised.

shane-huang commented on docs/readthedocs/source/doc/Nano/Howto/Training/TensorFlow/tensorflow_training_embedding_sparseadam.nblink:
```diff
@@ -0,0 +1,3 @@
+{
```
Embedding layers are often used to encode categorical items in deep learning applications. However, in applications such as recommendation systems, the embedding size may become huge due to large number of items or users, leading to extensive costs in both space and computation. 

Sparse Embedding is a common way to reduce the cost of very large embeddings. BigDL-Nano provides the Embedding layer class (with sparse support) and SparseAdam optimizer to optimize model with large embeddings in TensorFlow Keras applications.

Oscilloscope98 replied:
Have revised.
